pkg/cloud/services/elb/loadbalancer.go

```diff
@@ -218,6 +219,17 @@ func (s *Service) DeregisterInstanceFromAPIServerELB(i *infrav1.Instance) error
 	}
 
 	_, err = s.scope.ELB.DeregisterInstancesFromLoadBalancer(input)
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			case elb.ErrCodeAccessPointNotFoundException, elb.ErrCodeInvalidEndPointException:
```

Comment:
It looks like we are already doing a subset of this checking in [`IsNotFound`](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/212eae6835285a130aa1688a7ab565feb529b636/pkg/cloud/services/elb/errors.go#L58-L69), would it make sense to extend that function to also check for `elb.ErrCodeAccessPointNotFoundException` and use that here?

Reply:
@detiber The existing `IsNotFound` error does already check for `ErrCodeAccessPointNotFoundException`. I am additionally checking for `ErrCodeInvalidEndPointException` which would be an error thrown if the access point exists, but the instance you're trying to de-register was not.

So the change would be to also catch the `ErrCodeInvalidEndPointException` error in `IsNotFound` (which, during its other use in [`ReconcileLoadbalancers`](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/d4ade5922e04a2327f4043effd16252126dea223/pkg/cloud/services/elb/loadbalancer.go#L56) would probably never get returned). Happy to make that change though (catch that extra error code in `IsNotFound` and then simply call `IsNotFound` in `DeregisterInstanceFromAPIServerELB`.

Thinking about this more, I am realizing that this logic won't quite work and I am opening up a potential other issue with this change.
Just because we're de-registering an `AWSMachine` from the Load Balancer doesn't mean it is acceptable in **all** cases for the ELB to not exist. That condition is only acceptable if the `AWSCluster` itself is in a "delete reconcile" phase.
I'd hate to introduce some cross-check between different controllers, but do you also foresee a potential issue here that I should address? Or should I continue with the top level logic as is?

Reply:
I agree, `ErrCodeInvalidEndPointException` is semantically different, however the use case for `IsNotFound` is likely the same. It's unlikely that you need to do any cross-cluster reconciliation.

Might be better to keep this logic explicit here as you've done, in case there's a change to cluster reconciliation that would necessitate having to distinguish between the two cases. Probably long-term we should refactor the function names a bit better? WDYT @detiber?